### DISCUSSION TITLE
P-SMS-lib

### DISCUSSION THREAD

**Discussion by wrwrabbit** [wrwrabbit/Partisan-SMS#14] [General]
https://github.com/wrwrabbit/Partisan-Telegram-iOS/discussions/4

@vivabelarus

**Comment by billygates97** [comment]
@vivabelarus Привет.
Поресерчил еще на тему KMM, вы там отвергли либу Krypto
https://korlibs.soywiz.com/krypto/
Я так понимаю это готовый алгоритм шифрования под андпроид:
https://github.com/wrwrabbit/p-sms-lib/blob/master/src/androidMain/kotlin/by/cyberpartisan/psms/encryptor/AesEncryptor.kt
Так вот, в либе Krypto есть алгоритм шифрования AES CFB
Просто на странице Либы не написано почему-то)
https://github.com/korlibs/krypto/blob/743e9a2a8d0f2be36072a292c590f039acd133f5/krypto/src/commonMain/kotlin/com/soywiz/krypto/AES.kt
383 строчка
Или есть еще причины почему нам это не подходит?

**Reply by vivabelarus** [reply]
Добрый день. Окей, понял. Значит, можно попробовать её использовать. Больше проблем нет с KMM?

**Reply by billygates97** [reply]
@vivabelarus теперь все изи наверно.
Попробую сегодня сделать либу KMM и подрубить к IOS

**Reply by vivabelarus** [reply]
Да, отлично. Можно ещё заодно проверить, ловятся ли на свифте исключения, которые были выброшены в KMM.

**Reply by billygates97** [reply]
@vivabelarus короче целый день пытался сделать cocoa pod из библиотеки - не получилось, все перепробовал, не работает в KMM Это. Даже официальный пример.
Пришлось импортировать напрямую фреймворк.
Перевел всю либу на общий код, кроме кодировки 1251.
Я пока застрял, т.к. походу ее нет в KMM))
Либа для кодировок в KMM https://korlibs.soywiz.com/korio/
поддерживает только UTF8
Есть какие-то мысли? Вообще нам нужна эта кодировка?

**Reply by vivabelarus** [reply]
@billygates97 ну кодировку можно просто через when сделать. Там ж не очень много символов. Для теста можно вообще пока что скопировать реализацию UTF8.

**Reply by vivabelarus** [reply]
А платформенно-зависимая реализация всё-таки работает? Или только общий код работает?

**Reply by vivabelarus** [reply]
Если можно делать платформенно-зависимую реализацию, но она не может использовать фреймворк, то всё просто. Создаём в реализации функции обратного вызова, которые установит код на swift. По сути, swift внедрит необходимые зависимости. Это не обязательно должно быть реализовано как лямбда, можно и как класс, реализующий нужный интерфейс. Примерно так я себе это и представлял.

**Reply by billygates97** [reply]
@vivabelarus там я сделал для всех платформ что бы работало, т.е. общий код, а файлы для android я выпилил, т.к. они стали не нужны
Вообщем я протещу и залью с UTF8, я вам я предлагаю написать класс для перевода кодировок, там вроде не сложно 1251 маленькая кодировка, это гораздо лучше чем делать везде с платформы

**Reply by vivabelarus** [reply]
@billygates97, сделано

**Reply by billygates97** [reply]
@vivabelarus отлично, я потом смержу, у меня появилась проблема, KMM крашит IOS апку при кодировании/ раскодировании

let byteArray = KotlinByteArray(size: 0)
        let encryptedString = PSmsEncryptor().encode(str: message, key: byteArray, encryptionSchemeId: 0)
        print(PSmsEncryptor().decode(str: encryptedString, key: byteArray, encryptionSchemeId: 0))

Если сделать ключ больше, то крашит только при раскодировании, а кодирование норм происходит
Очень странно, я прогонял тесты в IDea, там все ок

**Reply by billygates97** [reply]
Uncaught Kotlin exception: by.cyberpartisan.psms.InvalidSignatureException

**Reply by billygates97** [reply]
Может есть идеи?

**Reply by vivabelarus** [reply]
Там при кодировании в результат зашивается хэш сообщения, при декодировании он проверяется. Делается это для того, чтобы случайно не декодировать левые сообщения, которые тоже закодированы в base64. Тут декодируется то же самое сообщение или просто случайный base64?

**Reply by billygates97** [reply]
@vivabelarus тоже самое, одна переменная

**Reply by vivabelarus** [reply]
А kmm можно при этом дебажить? Посмотреть, что он там делает

**Reply by billygates97** [reply]
Ну выводит что-то такое:
Function doesn't have or inherit @Throws annotation and thus exception isn't propagated from Kotlin to Objective-C/Swift as NSError.
It is considered unexpected and unhandled instead. Program will be terminated.
Uncaught Kotlin exception: kotlin.ArrayIndexOutOfBoundsException
    at 0   p_sms                               0x0000000102f23c60 ThrowArrayIndexOutOfBoundsException + 100
    at 1   p_sms                               0x0000000102faaac0 kfun:com.soywiz.krypto.AES#<init>(kotlin.ByteArray){} + 2700
    at 2   p_sms                               0x0000000102fac40c kfun:com.soywiz.krypto.AES.Companion#encryptAesCfb(kotlin.ByteArray;kotlin.ByteArray;kotlin.ByteArray;com.soywiz.krypto.Padding){}kotlin.ByteArray + 288
    at 3   p_sms                               0x0000000102fb14fc kfun:by.cyberpartisan.psms.encryptor.AesEncryptor#encrypt(kotlin.ByteArray;kotlin.ByteArray){}kotlin.ByteArray + 444

**Reply by billygates97** [reply]
@vivabelarus я завтра запушу, вы попробуете на андроиде, посмотрим все ли там ок

**Reply by vivabelarus** [reply]
Хм. Надо бы ещё посмотреть, почему вообще ругается на исключение.

**Reply by billygatesnew** [reply]
@vivabelarus перерегал аккаунт новый, пароли от старого чет не сохранились там где нужно, сорри
Можете, пожалуйста,  дать  доступ к репозиторию?

**Reply by vivabelarus** [reply]
@billygatesnew, пишите в бота киберов

**Reply by billygatesnew** [reply]
@vivabelarus запушил, тестаните плиз

**Reply by billygatesnew** [reply]
@vivabelarus сорри были проблемы с коммитами теперь перепушил)
